File: packages/guards/src/medical-advice.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

interface MedicalAdviceOptions {
  action: 'block' | 'warn';
}

const MEDICAL_PATTERNS: RegExp[] = [
  /\b(?:take|prescribe|recommend)\s+\d+\s*(?:mg|ml|mcg|units?)\b/gi,
  /\b(?:dosage|dose|prescription)\s*(?:of|:)\s*\d+/gi,
  /\b(?:diagnos(?:is|e|ed)|treatment for|cure for)\s+\w+/gi,
  /\byou (?:have|might have|probably have|could have)\s+(?:a\s+)?(?:\w+\s+){0,2}(?:disease|syndrome|disorder|condition|infection|cancer)\b/gi,
  /\bstop\s+taking\s+(?:your\s+)?(?:medication|medicine|pills)\b/gi,
  /\b(?:you\s+(?:should|need\s+to|must))\s+(?:get\s+)?(?:surgery|an?\s+(?:MRI|CT|X-ray|blood\s+test|biopsy))\b/gi,
  /\b(?:instead\s+of|rather\s+than)\s+(?:seeing|visiting|going\s+to)\s+(?:a\s+)?(?:doctor|physician|specialist)\b/gi,
];

export function medicalAdvice(options: MedicalAdviceOptions): Guard {
  return {
    name: 'medical-advice',
    version: '0.1.0',
    description: 'Detect unauthorized medical advice in LLM responses',
    category: 'content',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];
      for (const p of MEDICAL_PATTERNS) {
        const re = new RegExp(p.source, p.flags);
        const m = re.exec(text);
        if (m) matched.push(m[0]);
      }
      const triggered = matched.length > 0;
      return {
        guardName: 'medical-advice',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        message: triggered ? `Medical advice detected: "${matched[0]}"` : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matched, reason: 'LLM should not provide specific medical advice without disclaimers' } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/medical_advice.py
"""Detects medical advice patterns in text."""
from __future__ import annotations

import re
import time
from typing import List

from open_guardrail.core import GuardResult

_PATTERNS = [
    re.compile(r"take\s+\w+\s+\d+\s*mg", re.IGNORECASE),
    re.compile(r"you\s+should\s+see\s+a\s+doctor", re.IGNORECASE),
    re.compile(r"diagnosis\s+is\b", re.IGNORECASE),
    re.compile(r"treatment\s+plan\b", re.IGNORECASE),
    re.compile(r"\bprescribe\b", re.IGNORECASE),
]


class _MedicalAdvice:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "medical-advice"
        self.action = action

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matched: List[str] = []
        for p in _PATTERNS:
            m = p.search(text)
            if m:
                matched.append(m.group())
        triggered = len(matched) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="medical-advice",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"Medical advice detected: {len(matched)} pattern(s)" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"matched": matched} if triggered else None,
        )


def medical_advice(*, action: str = "warn") -> _MedicalAdvice:
    return _MedicalAdvice(action=action)
